src/lib/optimizer/strategy/semi_join_reduction_rule.cpp

```diff
@@ -0,0 +1,140 @@
+        lqp_insert_node(join_node, side_of_join, semi_join_reduction_node);
+
+        semi_join_reduction_node->set_right_input(right_input);
+
+        const auto reduction_input_cardinality =
+            estimator->estimate_cardinality(semi_join_reduction_node->left_input());
+        const auto reduction_output_cardinality = estimator->estimate_cardinality(semi_join_reduction_node);
+
+        semi_join_reduction_node->set_right_input(nullptr);
+        lqp_remove_node(semi_join_reduction_node);
+
+        if (reduction_input_cardinality == 0 ||
+            reduction_output_cardinality / reduction_input_cardinality > MINIMUM_SELECTIVITY) {
+          return;
+        }
+
+        // For `t1 JOIN t2 on t1.a = t2.a` where a semi join reduction is supposed to be added to the t1, `t1.a = t2.a`
+        // is the predicate_expression. The values from t1 are supposed to be filtered by looking at t1.a, which is
+        // called the left_expression.
+        const auto& left_expression = expression_evaluable_on_lqp(predicate_expression->left_operand(), *right_input)
+                                          ? predicate_expression->left_operand()
+                                          : predicate_expression->right_operand();
+        DebugAssert(!expression_evaluable_on_lqp(left_expression, *join_node->input(side_of_join)),
+                    "Expected filtered expression to be uniquely evaluable on one side of the join");
+
+        // Currently, the right input of the semi join reduction is opposite_side(side_of_join). However, on that side,
+        // multiple joins have been added that increase the cardinality. We walk the LQP starting with right_input
+        // taking those paths where lext_expression continues to be evaluable.
+        while (true) {
+          if (right_input->type != LQPNodeType::Join) {
+            break;
+          }
+          // No matter the join type, we will never see values that were not there before
+
+          const auto try_deeper_right_input = [&](const auto side_of_input) {
+            const auto& candidate = right_input->input(side_of_input);
+            if (!expression_evaluable_on_lqp(left_expression, *candidate)) {
+              return false;
+            }
+
+            const auto candidate_cardinality = estimator->estimate_cardinality(candidate);
+
+            // Check if the candidate has an equal or lower cardinality. Allow for some tolerance due to float values
+            // being used.
+            if (candidate_cardinality > right_cardinality * 1.01) {
+              return false;
+            }
+
+            right_input = candidate;
+            right_cardinality = candidate_cardinality;
+
+            return true;
+          };
+
+          if (try_deeper_right_input(LQPInputSide::Left)) continue;
+          if (try_deeper_right_input(LQPInputSide::Right)) continue;
+          break;
+        }
+
+        if (right_cardinality > left_cardinality) {
```

Comment:
I thought about that, but I saw some value in TPC-x plans looking similar independent of the scale factor

Reply:
Well, I can understand that from a debugging view point but not from the stand point that the optimizer should produce the best possible plan for the given current system.

Reply:
How would we come up with a meaningful threshold for that? This depends on data types, cardinalities, selectivity, ...

Reply:
I would be "as usual" a rough estimate of a hash table size that might fit into L1. Could be very rough, but most tresholds are and it could still handle many cases. It probably does not make a huge difference if it's 1,000 or 500, but that might already be enough to cover almost all cases where a semi hash join with a bigger build relation is still beneficial.

I see your point that this estimate is anything but accurate, but then again: which current threshold in the optimizer are not?

Reply:
Thinking about it a little bit longer, I don't think the threshold is needed here. The input where this comes into play it so small that the overhead of the added semi join is probably not worth it anyways.
Resolving.